From 4b5a3a75f9df7bc03db24fc71929d695c708eb01 Mon Sep 17 00:00:00 2001
From: seven332 <seven332@163.com>
Date: Fri, 23 Dec 2016 01:15:29 +0800
Subject: [PATCH] Add decoding cmyk to rgb

---
 jdcolext.c | 39 +++++++++++++++++++++++++++++++++++
 jdcolor.c  | 60 ++++++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 99 insertions(+)

diff --git a/jdcolext.c b/jdcolext.c
index 59b676cc4..6a38d2ab0 100644
--- a/jdcolext.c
+++ b/jdcolext.c
@@ -141,3 +141,42 @@ rgb_rgb_convert_internal (j_decompress_ptr cinfo,
     }
   }
 }
+
+
+/*
+ * Convert CMYK to extended RGB
+ */
+
+INLINE
+LOCAL(void)
+cmyk_rgb_convert_internal (j_decompress_ptr cinfo,
+                           JSAMPIMAGE input_buf, JDIMENSION input_row,
+                           JSAMPARRAY output_buf, int num_rows)
+{
+  register int k;
+  register JSAMPROW inptr0, inptr1, inptr2, inptr3;
+  register JSAMPROW outptr;
+  register JDIMENSION col;
+  JDIMENSION num_cols = cinfo->output_width;
+
+  while (--num_rows >= 0) {
+    inptr0 = input_buf[0][input_row];
+    inptr1 = input_buf[1][input_row];
+    inptr2 = input_buf[2][input_row];
+    inptr3 = input_buf[3][input_row];
+    input_row++;
+    outptr = *output_buf++;
+    for (col = 0; col < num_cols; col++) {
+      k = GETJSAMPLE(inptr3[col]);
+      outptr[RGB_RED] = (JSAMPLE) (GETJSAMPLE(inptr0[col]) * k / MAXJSAMPLE);
+      outptr[RGB_GREEN] = (JSAMPLE) (GETJSAMPLE(inptr1[col]) * k / MAXJSAMPLE);
+      outptr[RGB_BLUE] = (JSAMPLE) (GETJSAMPLE(inptr2[col]) * k / MAXJSAMPLE);
+      /* Set unused byte to 0xFF so it can be interpreted as an opaque */
+      /* alpha channel value */
+#ifdef RGB_ALPHA
+      outptr[RGB_ALPHA] = 0xFF;
+#endif
+      outptr += RGB_PIXELSIZE;
+    }
+  }
+}
diff --git a/jdcolor.c b/jdcolor.c
index ab8fa2492..5f22715a9 100644
--- a/jdcolor.c
+++ b/jdcolor.c
@@ -105,6 +105,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #define ycc_rgb_convert_internal  ycc_extrgb_convert_internal
 #define gray_rgb_convert_internal  gray_extrgb_convert_internal
 #define rgb_rgb_convert_internal  rgb_extrgb_convert_internal
+#define cmyk_rgb_convert_internal  cmyk_extrgb_convert_internal
 #include "jdcolext.c"
 #undef RGB_RED
 #undef RGB_GREEN
@@ -113,6 +114,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #undef ycc_rgb_convert_internal
 #undef gray_rgb_convert_internal
 #undef rgb_rgb_convert_internal
+#undef cmyk_rgb_convert_internal
 
 #define RGB_RED  EXT_RGBX_RED
 #define RGB_GREEN  EXT_RGBX_GREEN
@@ -122,6 +124,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #define ycc_rgb_convert_internal  ycc_extrgbx_convert_internal
 #define gray_rgb_convert_internal  gray_extrgbx_convert_internal
 #define rgb_rgb_convert_internal  rgb_extrgbx_convert_internal
+#define cmyk_rgb_convert_internal  cmyk_extrgbx_convert_internal
 #include "jdcolext.c"
 #undef RGB_RED
 #undef RGB_GREEN
@@ -131,6 +134,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #undef ycc_rgb_convert_internal
 #undef gray_rgb_convert_internal
 #undef rgb_rgb_convert_internal
+#undef cmyk_rgb_convert_internal
 
 #define RGB_RED  EXT_BGR_RED
 #define RGB_GREEN  EXT_BGR_GREEN
@@ -139,6 +143,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #define ycc_rgb_convert_internal  ycc_extbgr_convert_internal
 #define gray_rgb_convert_internal  gray_extbgr_convert_internal
 #define rgb_rgb_convert_internal  rgb_extbgr_convert_internal
+#define cmyk_rgb_convert_internal  cmyk_extbgr_convert_internal
 #include "jdcolext.c"
 #undef RGB_RED
 #undef RGB_GREEN
@@ -147,6 +152,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #undef ycc_rgb_convert_internal
 #undef gray_rgb_convert_internal
 #undef rgb_rgb_convert_internal
+#undef cmyk_rgb_convert_internal
 
 #define RGB_RED  EXT_BGRX_RED
 #define RGB_GREEN  EXT_BGRX_GREEN
@@ -156,6 +162,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #define ycc_rgb_convert_internal  ycc_extbgrx_convert_internal
 #define gray_rgb_convert_internal  gray_extbgrx_convert_internal
 #define rgb_rgb_convert_internal  rgb_extbgrx_convert_internal
+#define cmyk_rgb_convert_internal  cmyk_extbgrx_convert_internal
 #include "jdcolext.c"
 #undef RGB_RED
 #undef RGB_GREEN
@@ -165,6 +172,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #undef ycc_rgb_convert_internal
 #undef gray_rgb_convert_internal
 #undef rgb_rgb_convert_internal
+#undef cmyk_rgb_convert_internal
 
 #define RGB_RED  EXT_XBGR_RED
 #define RGB_GREEN  EXT_XBGR_GREEN
@@ -174,6 +182,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #define ycc_rgb_convert_internal  ycc_extxbgr_convert_internal
 #define gray_rgb_convert_internal  gray_extxbgr_convert_internal
 #define rgb_rgb_convert_internal  rgb_extxbgr_convert_internal
+#define cmyk_rgb_convert_internal  cmyk_extxbgr_convert_internal
 #include "jdcolext.c"
 #undef RGB_RED
 #undef RGB_GREEN
@@ -183,6 +192,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #undef ycc_rgb_convert_internal
 #undef gray_rgb_convert_internal
 #undef rgb_rgb_convert_internal
+#undef cmyk_rgb_convert_internal
 
 #define RGB_RED  EXT_XRGB_RED
 #define RGB_GREEN  EXT_XRGB_GREEN
@@ -192,6 +202,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #define ycc_rgb_convert_internal  ycc_extxrgb_convert_internal
 #define gray_rgb_convert_internal  gray_extxrgb_convert_internal
 #define rgb_rgb_convert_internal  rgb_extxrgb_convert_internal
+#define cmyk_rgb_convert_internal  cmyk_extxrgb_convert_internal
 #include "jdcolext.c"
 #undef RGB_RED
 #undef RGB_GREEN
@@ -201,6 +212,7 @@ typedef my_color_deconverter *my_cconvert_ptr;
 #undef ycc_rgb_convert_internal
 #undef gray_rgb_convert_internal
 #undef rgb_rgb_convert_internal
+#undef cmyk_rgb_convert_internal
 
 
 /*
@@ -524,6 +536,52 @@ rgb_rgb_convert (j_decompress_ptr cinfo,
 }
 
 
+/*
+ * Convert CMYK to RGB
+ */
+
+METHODDEF(void)
+cmyk_rgb_convert (j_decompress_ptr cinfo,
+                  JSAMPIMAGE input_buf, JDIMENSION input_row,
+                  JSAMPARRAY output_buf, int num_rows)
+{
+  switch (cinfo->out_color_space) {
+    case JCS_EXT_RGB:
+      cmyk_extrgb_convert_internal(cinfo, input_buf, input_row, output_buf,
+                                   num_rows);
+          break;
+    case JCS_EXT_RGBX:
+    case JCS_EXT_RGBA:
+      cmyk_extrgbx_convert_internal(cinfo, input_buf, input_row, output_buf,
+                                    num_rows);
+          break;
+    case JCS_EXT_BGR:
+      cmyk_extbgr_convert_internal(cinfo, input_buf, input_row, output_buf,
+                                   num_rows);
+          break;
+    case JCS_EXT_BGRX:
+    case JCS_EXT_BGRA:
+      cmyk_extbgrx_convert_internal(cinfo, input_buf, input_row, output_buf,
+                                    num_rows);
+          break;
+    case JCS_EXT_XBGR:
+    case JCS_EXT_ABGR:
+      cmyk_extxbgr_convert_internal(cinfo, input_buf, input_row, output_buf,
+                                    num_rows);
+          break;
+    case JCS_EXT_XRGB:
+    case JCS_EXT_ARGB:
+      cmyk_extxrgb_convert_internal(cinfo, input_buf, input_row, output_buf,
+                                    num_rows);
+          break;
+    default:
+      cmyk_rgb_convert_internal(cinfo, input_buf, input_row, output_buf,
+                                num_rows);
+          break;
+  }
+}
+
+
 /*
  * Adobe-style YCCK->CMYK conversion.
  * We convert YCbCr to R=1-C, G=1-M, and B=1-Y using the same
@@ -835,6 +893,8 @@ jinit_color_deconverter (j_decompress_ptr cinfo)
         cconvert->pub.color_convert = null_convert;
       else
         cconvert->pub.color_convert = rgb_rgb_convert;
+    } else if (cinfo->jpeg_color_space == JCS_CMYK) {
+      cconvert->pub.color_convert = cmyk_rgb_convert;
     } else
       ERREXIT(cinfo, JERR_CONVERSION_NOTIMPL);
     break;
